### PR TITLE
statistics: handle deleted tables correctly in the PQ

### PR DESCRIPTION
**Pull request by Rustin170506**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/57648

Problem Summary:

### What changed and how does it work?

1. Correctly skip the deleted table. Sorry for overlooking this.
2. Make sure we don't write to nil runningJobs and mustRetryJobs because the worker might access them again after the queue is closed (we don't stop the worker at the same time because we don't want to block the etcd election process). Accessing it is ok because we will reinitialize the queue next time. So we will take care of the failed jobs next time.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/statistics/handle/autoanalyze/priorityqueue/analysis_job_factory.go [hunk 1]
```diff
@@ -57,7 +57,7 @@ func (f *AnalysisJobFactory) CreateNonPartitionedTableAnalysisJob(
 	tblInfo *model.TableInfo,
 	tblStats *statistics.Table,
 ) AnalysisJob {
-	if !tblStats.IsEligibleForAnalysis() {
+	if tblStats == nil || !tblStats.IsEligibleForAnalysis() {
```
Comment (Rustin170506) [inline pkg/statistics/handle/autoanalyze/priorityqueue/analysis_job_factory.go:60]: We didn't encounter any issues here, just double-checking.

File: pkg/statistics/handle/autoanalyze/priorityqueue/analysis_job_factory.go [hunk 2]
```diff
@@ -92,7 +92,7 @@ func (f *AnalysisJobFactory) CreateStaticPartitionAnalysisJob(
 	partitionID int64,
 	partitionStats *statistics.Table,
 ) AnalysisJob {
-	if !partitionStats.IsEligibleForAnalysis() {
+	if partitionStats == nil || !partitionStats.IsEligibleForAnalysis() {
```
Comment (Rustin170506) [inline pkg/statistics/handle/autoanalyze/priorityqueue/analysis_job_factory.go:95]: ditto

File: pkg/statistics/handle/autoanalyze/priorityqueue/analysis_job_factory.go [hunk 3]
```diff
@@ -128,7 +128,7 @@ func (f *AnalysisJobFactory) CreateDynamicPartitionedTableAnalysisJob(
 	globalTblStats *statistics.Table,
 	partitionStats map[PartitionIDAndName]*statistics.Table,
 ) AnalysisJob {
-	if !globalTblStats.IsEligibleForAnalysis() {
+	if globalTblStats == nil || !globalTblStats.IsEligibleForAnalysis() {
```
Comment (Rustin170506) [inline pkg/statistics/handle/autoanalyze/priorityqueue/analysis_job_factory.go:131]: ditto

File: pkg/statistics/handle/autoanalyze/priorityqueue/queue.go [hunk 4]
```diff
@@ -852,9 +854,9 @@ func (pq *AnalysisPriorityQueue) Close() {
 	pq.syncFields.initialized = false
 	// The rest fields will be reset when the priority queue is initialized.
 	// But we do it here for double safety.
-	pq.syncFields.inner = nil
-	pq.syncFields.runningJobs = nil
-	pq.syncFields.mustRetryJobs = nil
+	pq.syncFields.inner = newHeap()
```
Comment (Rustin170506) [inline pkg/statistics/handle/autoanalyze/priorityqueue/queue.go:857]: Fix the `[category=stats] [recover="assignment to entry in nil map"]`

File: pkg/statistics/handle/autoanalyze/priorityqueue/queue.go [hunk 5]
```diff
@@ -644,14 +644,16 @@ func (pq *AnalysisPriorityQueue) RefreshLastAnalysisDuration() {
...
+				// Delete the job from the queue since its table is missing. This is a safeguard -
+				// DDL events should have already cleaned up jobs for dropped tables.
 				err := pq.syncFields.inner.delete(job)
 				if err != nil {
 					statslogutil.StatsLogger().Error("Failed to delete job from priority queue",
 						zap.Error(err),
 						zap.String("job", job.String()),
 					)
 				}
+				continue
```
Comment (Rustin170506) [inline pkg/statistics/handle/autoanalyze/priorityqueue/queue.go:656]: Fix the `[category=stats] [recover="runtime error: invalid memory address or nil pointer dereference"]`

File: pkg/statistics/handle/autoanalyze/priorityqueue/queue_test.go [hunk 6]
```diff
@@ -608,3 +609,47 @@ func TestPQCanBeClosedAndReInitialized(t *testing.T) {
 	// Check if the priority queue is initialized.
 	require.True(t, pq.IsInitialized())
 }
+
+func TestPQHandlesTableDeletionGracefully(t *testing.T) {
```
Comment (Rustin170506) [inline pkg/statistics/handle/autoanalyze/priorityqueue/queue_test.go:613]: Test case for: `[category=stats] [recover="runtime error: invalid memory address or nil pointer dereference"]`

**Review by Rustin170506 — COMMENTED** [review]
🔢 Self-check (PR reviewed by myself and ready for feedback.)

File: pkg/statistics/handle/autoanalyze/priorityqueue/queue.go [hunk 7]
```diff
@@ -752,11 +754,24 @@ func (pq *AnalysisPriorityQueue) Pop() (AnalysisJob, error) {
...
 		delete(pq.syncFields.runningJobs, j.GetTableID())
 	})
 	job.RegisterFailureHook(func(j AnalysisJob, needRetry bool) {
 		pq.syncFields.mu.Lock()
 		defer pq.syncFields.mu.Unlock()
+		// During owner switch, the priority queue is closed and its fields are reset to nil.
+		// We allow running jobs to complete normally rather than stopping them, so this nil check
+		// is expected when jobs finish after the switch. Failed jobs will be handled by the next
+		// initialization, so we can safely ignore them here.
+		if pq.syncFields.runningJobs == nil || pq.syncFields.mustRetryJobs == nil {
```
Comment (Rustin170506) [inline pkg/statistics/handle/autoanalyze/priorityqueue/queue.go:772]: For the ` ["Auto analyze job execution failed"] [category=stats] [recover="assignment to entry in nil map"]`.

**Review by Rustin170506 — COMMENTED** [review]
🔢 Self-check (PR reviewed by myself and ready for feedback.)

**Review by qw4990 — APPROVED** [review]

**Review by elsa0520 — APPROVED** [review]
LGTM

**Pull request merged**